server/statistics/hot_peer_cache.go

```diff
@@ -108,120 +112,93 @@ func (f *hotPeerCache) Update(item *HotPeerStat) {
+	storeID := peer.GetStoreID()
+	deltaLoads := f.getFlowDeltaLoads(peer)
 	loads := make([]float64, len(deltaLoads))
 	for i := range deltaLoads {
 		loads[i] = deltaLoads[i] / float64(interval)
 	}
-	f.collectRegionMetrics(loads, interval)
-
-	// old region is in the front and new region is in the back
-	// which ensures it will hit the cache if moving peer or transfer leader occurs with the same replica number
-
+	justTransferLeader := f.justTransferLeader(region)
+	// transfer read leader or remove write peer
+	isExpired := f.isPeerExpired(peer, region)
+	oldItem := f.getOldHotPeerStat(region.GetID(), storeID)
+	if !isExpired && Denoising && interval < HotRegionReportMinInterval {
+		return nil
+	}
+	thresholds := f.calcHotThresholds(storeID)
 	var peers []uint64
 	for _, peer := range region.GetPeers() {
 		peers = append(peers, peer.StoreId)
 	}
-
-	var tmpItem *HotPeerStat
-	storeIDs := f.getAllStoreIDs(region)
-	justTransferLeader := f.justTransferLeader(region)
-	for _, storeID := range storeIDs {
-		isExpired := f.isRegionExpired(region, storeID) // transfer read leader or remove write peer
-		oldItem := f.getOldHotPeerStat(region.GetID(), storeID)
-		if isExpired && oldItem != nil { // it may has been moved to other store, we save it to tmpItem
-			tmpItem = oldItem
-		}
-
-		// This is used for the simulator and test. Ignore if report too fast.
-		if !isExpired && Denoising && interval < HotRegionReportMinInterval {
-			continue
-		}
-
-		thresholds := f.calcHotThresholds(storeID)
-
-		newItem := &HotPeerStat{
-			StoreID:            storeID,
-			RegionID:           region.GetID(),
-			Kind:               f.kind,
-			Loads:              loads,
-			LastUpdateTime:     time.Now(),
-			needDelete:         isExpired,
-			isLeader:           region.GetLeader().GetStoreId() == storeID,
-			justTransferLeader: justTransferLeader,
-			interval:           interval,
-			peers:              peers,
-			thresholds:         thresholds,
-		}
-
-		if oldItem == nil {
-			if tmpItem != nil { // use the tmpItem cached from the store where this region was in before
-				oldItem = tmpItem
-			} else { // new item is new peer after adding replica
-				for _, storeID := range storeIDs {
-					oldItem = f.getOldHotPeerStat(region.GetID(), storeID)
-					if oldItem != nil {
-						break
-					}
-				}
+	newItem := &HotPeerStat{
+		StoreID:            storeID,
+		RegionID:           region.GetID(),
+		Kind:               f.kind,
+		Loads:              loads,
+		LastUpdateTime:     time.Now(),
+		needDelete:         isExpired,
+		isLeader:           region.GetLeader().GetStoreId() == storeID,
+		justTransferLeader: justTransferLeader,
+		interval:           interval,
+		peers:              peers,
+		thresholds:         thresholds,
+	}
+	if oldItem == nil {
+		for _, storeID := range f.getAllStoreIDs(region) {
+			oldItem = f.getOldHotPeerStat(region.GetID(), storeID)
+			if oldItem != nil {
+				break
```

Comment:
The behavior is different. The previous code ensured that it inherited exactly the right information by saving the expired item as a tmpItem. The current code is equivalent to inheriting a random item.

Reply:
Currently, we tend to inherit stat from leader, otherwise we inherit stat from random peer. In future, we will use a new way to inherit stat more currectly.